### PR TITLE
Feat navsys astar

### PR DESCRIPTION
**Pull request by LuukSteeman**
A* algorithm implemented
Added a pathnode class which adds some extra functionality needed for the algorithm, more on that can be found in the in-code documentation
Added some extra functionality to vertice and node classes in the form of get() functions.

### CODE REVIEW

**Comment by philippezwietering** [comment]
Two compiler warnings emerged, not ready for merge
`/home/philippe/Documents/School/hu-r2d2/R2D2-2017/modules/NAVSYS/src/pathnode.cc:14:5: warning: field 'coordinate' will be initialized after field 'g' [-Wreorder]
    coordinate(n.getCoordinate()),
    ^
In file included from /home/philippe/Documents/School/hu-r2d2/R2D2-2017/modules/NAVSYS/src/pathnode.cc:8:
/home/philippe/Documents/School/hu-r2d2/R2D2-2017/modules/NAVSYS/src/pathnode.hh:34:11: warning: private field 'h' is not used [-Wunused-private-field]
    float h;
          ^
2 warnings generated.
`

File: modules/NAVSYS/src/astar.hh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#include <algorithm>
+#include <limits>
+#include <queue>
+#include "coordinate.hh"
+#include "graph.hh"
+#include "node.hh"
+#include "pathnode.hh"
+#include "vertice.hh"
+
+using namespace std;
```
Comment (sqrtroot) [inline modules/NAVSYS/src/astar.hh]: No `using namespace` http://llvm.org/docs/CodingStandards.html#do-not-use-using-namespace-std

File: modules/NAVSYS/src/astar.cc [hunk 2]
```diff
@@ -0,0 +1,189 @@
+#include "astar.hh"
+
+vector<PathNode> aStar(Graph * g, Node *start, Node *goal)
```
Comment (sqrtroot) [inline modules/NAVSYS/src/astar.cc]: Why use raw pointers here and save pointers in the rest of the code. Raw pointers might not be save.
Couldn't you pass them by reference or const reference?

File: modules/NAVSYS/src/pathnode.hh [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+class PathNode : public Node
+{
+private:
+    
+    /**
+    * The total vertice weight from the start node to this node
+    * g is the total weight of all vertices passed when going to this node from the starting node
+    * g changes depending on what path is taken towards this node
+    */
+    float g;
```
Comment (sqrtroot) [inline modules/NAVSYS/src/pathnode.hh]: name is not a noun/does not give any info
http://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly (also look at the wiki as we diverge from the standard a little bit here)

File: modules/NAVSYS/src/pathnode.hh [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    /**
+    * The euclidian (as the crow flies) distance to the end node
+    * h is the straight distance between this node and the end node, calculated with the pythagorian theorem
+    * In this class, it's calculated by calling the euclideanDistance(const Node &other) function from node
+    * Which in turn calls the euclideanDistance(const Coordinate &rhs) function of coordinate
+    */
+    float h;
```
Comment (sqrtroot) [inline modules/NAVSYS/src/pathnode.hh]: same as `g`

File: modules/NAVSYS/src/pathnode.hh [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    /**
+    * f is the priority of this node, calculated with f = g + h
+    * f is used to determine how high a nodes priority is, a lower f is a higher priority
+    * f is calculated on object construction and when the parent is set
+    * This makes sure that f is always correct
+    */
+    float f = 0;
```
Comment (sqrtroot) [inline modules/NAVSYS/src/pathnode.hh]: same as `g` & `h`

**Review by sqrtroot — CHANGES_REQUESTED** [review]
See comments in code

**Comment by LuukSteeman** [comment]
Those variables are standard nomenclature for A*, do you still want me to change them???

File: modules/NAVSYS/src/pathnode.hh [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    /**
+    * coordinate is the PathNodes location
+    */
+    Coordinate coordinate;
```
Comment (sqrtroot) [inline modules/NAVSYS/src/pathnode.hh]: You already inherit this from Node

Comment (LuukSteeman) [reply modules/NAVSYS/src/pathnode.hh]: if i tried that it started complaining Node doesn't have a default constructor

File: modules/NAVSYS/src/astar.hh (hunk 1 above)
Comment (LuukSteeman) [reply modules/NAVSYS/src/astar.hh]: well that sucks :(

File: modules/template-arduino/CMakeLists.txt [hunk 7]
```diff
@@ -1,29 +1,36 @@
 cmake_minimum_required (VERSION 2.8.11)
```
Comment (sqrtroot) [inline modules/template-arduino/CMakeLists.txt]: You did `git add *` didn't you :no_good_man:
Please remove changes from this file

File: modules/NAVSYS/src/pathnode.hh (hunk 6 above)
Comment (sqrtroot) [reply modules/NAVSYS/src/pathnode.hh]: You have to call your parent classes constructor so it gets initialized and doesn't try to automagically do it

**Comment by sqrtroot** [comment]
For someone that doesn't know what a star does seeing those variables in code may not give them any info without looking at the doxygen. It would be beneficial if you change them so the code makes more sense for people without a star knowledge

File: modules/NAVSYS/src/pathnode.hh (hunk 6 above)
Comment (LuukSteeman) [reply modules/NAVSYS/src/pathnode.hh]: fixed

File: modules/NAVSYS/src/pathnode.hh (hunk 4 above)
Comment (LuukSteeman) [reply modules/NAVSYS/src/pathnode.hh]: removed h as it isn't used anyway

**Comment by LuukSteeman** [comment]
will do

File: modules/NAVSYS/src/pathnode.hh (hunk 5 above)
Comment (LuukSteeman) [reply modules/NAVSYS/src/pathnode.hh]: changed for f and g

**Comment by LuukSteeman** [comment]
@philippezwietering i fixed those warnings

File: modules/NAVSYS/src/astar.cc (hunk 2 above)
Comment (LuukSteeman) [reply modules/NAVSYS/src/astar.cc]: it breaks the linker when i try that:
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2019	unresolved external symbol "class std::vector<class PathNode,class std::allocator<class PathNode> > __cdecl aStar(class Graph const &,class Node const &,class Node const &)" (?aStar@@YA?AV?$vector@VPathNode@@V?$allocator@VPathNode@@@std@@@std@@ABVGraph@@ABVNode@@1@Z) referenced in function _main	navsys	
\modules\NAVSYS\build\main.obj	1

File: modules/NAVSYS/src/astar.hh (hunk 1 above)
Comment (LuukSteeman) [reply modules/NAVSYS/src/astar.hh]: changed

**Review by sqrtroot — APPROVED** [review]

**Comment by LuukSteeman** [comment]
[Luuk Steeman](https://trello.com/c/pdYmMpYR/8-luuk-steeman)

**Pull request merged**